### DISCUSSION TITLE
Hybrid searches are slower

### DISCUSSION THREAD

**Discussion by YangZhaoo** [milvus-io/milvus#29519] [Q&A and General discussion] [answered]
Hybrid searches are slower. 
Approximately 60 vectors are searched for similar vectors at a time. After creating the index on the scalar field, the end-to-end delay dropped from 1.7 s to 1.2 seconds. My application and milvus are both on the same intranet. 
Is this normal? Thank you

**Comment by yhmo** [comment]
How many entities are in the collection?

**Reply by YangZhaoo** [reply]
total 296,685,    after scalar field filter maybe has 10,000 entities

**Reply by yhmo** [reply]
The scalar field for filtering is integer or varchar?

**Reply by YangZhaoo** [reply]
boolean expression like this _user_id == "123456" and voice_id == 123_, both integer and varchar.

**Reply by yhmo** [reply]
I wrote a script to test. 300000 entities, IVF_FLAT index for vector field. I didn't create index for the scalar fields.
The average search latency is no more than 100ms.

```
import random
import time

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)


connections.connect(host='localhost', port='19530')
print(utility.get_server_version())

collection_name = "test"
dim = 1024
metric_type = "L2"

# create collection
fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
    FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
    FieldSchema(name="name", dtype = DataType.VARCHAR, max_length=1024),
    FieldSchema(name="age", dtype = DataType.INT64),
]

schema = CollectionSchema(fields)

if utility.has_collection(collection_name):
    utility.drop_collection(collection_name)

collection = Collection(name=collection_name, schema=schema)
print(f"Collection '{collection_name}' created")

# insert
batch_size = 10000
for i in range(30):
    print("insert batch", i)
    data = [
        [[random.random() for _ in range(dim)] for _ in range(batch_size)],
        [f"name_{k%10}" for k in range(batch_size)], # name
        [k%100 for k in range(batch_size)],  # age
    ]
    collection.insert(data)

collection.flush()
print(collection.num_entities)

index_params = {
    'metric_type': metric_type,
    'index_type': "IVF_FLAT",
    'params': {"nlist": 512},
}
collection.create_index(field_name="vector", index_params=index_params)
print("index created")

collection.load()
print("collection loaded")

for i in range(10):
    print("search...")
    start = time.time()
    results = collection.search(
        data=[[random.random() for _ in range(dim)] for _ in range(60)],
        anns_field="vector",
        expr="name==\"name_3\" and age > 50",
        param={
            "metric_type": metric_type,
            "params": {"nprobe":32}
        },
        output_fields=["id", "name", "age"],
        consistency_level="Bounded",
        limit=10,
    )
    end = time.time()
    print(f"search latency: {(end-start)*1000}ms")

    # for hits in results:
    #     print(f"results of target vector", len(hits))
    #     for hit in hits:
    #         print(hit.to_dict())

```

**Reply by YangZhaoo** [reply]
Sorry, I just forgot to mention it. To pursue 100% recall rate, we used FLAT, not IVF_FLAT, the vector dimension is 1024, float type. I don't know if this has any impact on search performance.
![image](https://github.com/milvus-io/milvus/assets/41366656/d1417559-407c-46e9-93b4-33b9e53eab22)
collection schema like this:
![image](https://github.com/milvus-io/milvus/assets/41366656/d927b056-af03-4db1-9404-c06032594d54)

**Reply by YangZhaoo** [reply]
code like this
![image](https://github.com/milvus-io/milvus/assets/41366656/3fa2c96b-8dc9-416a-86dc-aa4206b997ff)

**Reply by YangZhaoo** [reply]
At present topK is 1

**Reply by yhmo** [reply, marked as answer]
withConsistencyLevel(ConsistencyLevelEnum.STRONG)   =======change to=======> withConsistencyLevel(ConsistencyLevelEnum.BOUNDED) 


STRONG consistency level will add an extra 200ms~400ms to search latency.

**Reply by yhmo** [reply]
With FLAT index, the latency depends on how many entities are filtered out to do ANN search.
If you set an empty expression, the latency to brute-force search the 300000 entities is about 2.5 seconds. But if you set an expression, and 10000 entities match the expression, latency is about 200ms.

**Reply by YangZhaoo** [reply]
I change consistentcyLevel from STRONG to BOUNDED, The latency of querying a vector is reduced approximate 400ms.
Thank you very much.

**Reply by YangZhaoo** [reply]
milvus version: v2.3.2-dev
hello, I used milvus cluster with cpu. queryNode has 3 instances.
with FLAT index, set an boolean expression and 57329 entities match the expression. 
the vector dimension is 1024.
Approximately 60 vectors are searched for similar vectors at a time and latency is about 1900ms.
Is this normal? Are there any configuration items that can be optimized?
@yhmo Thank you very much.

**Reply by yhmo** [reply]
1 million entities, 1024dim, FLAT, 50000 entities match the expression. 60 vectors to search
The average time cost is 300~400ms on my machine with a standalone milvus:

The test script:
```
import random
import time
import json
import numpy as np

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)

connections.connect(host='localhost', port='19530')
print(utility.get_server_version())


collection_name = "test"
dim = 1024
metric_type = "L2"


fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
    FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
    FieldSchema(name="name", dtype = DataType.VARCHAR, max_length=1024),
]

schema = CollectionSchema(fields=fields, enable_dynamic_field=False)

if utility.has_collection(collection_name):
    utility.drop_collection(collection_name)

collection = Collection(name=collection_name, schema=schema)
print(f"Collection '{collection_name}' created")


# insert
batch_size = 10000
for i in range(100):
    print("insert batch", i)
    data = [
        [[random.random() for _ in range(dim)] for _ in range(batch_size)],
        [f"name_{k%20}" for k in range(batch_size)], # name
    ]
    collection.insert(data)

collection.flush()

index_params = {
    'metric_type': metric_type,
    'index_type': "FLAT",
    'params': {},
}
collection.create_index(field_name="vector", index_params=index_params)
print("index created")


collection = Collection(collection_name)
collection.load()
print("collection loaded")

res = collection.query(expr="", consistency_level="Strong", output_fields=["count(*)"])
print("count(*):", res[0]["count(*)"])
print("collection.num_entities = ", collection.num_entities)


def search(params):
    print("=================== search ==================")
    start = time.time()
    results = collection.search(
        expr="name like 'name_4%'",
        data=[[random.random() for _ in range(dim)] for _ in range(60)],
        anns_field="vector",
        param=params,
        limit=10,
    )
    # for hits in results:
    #     print(f"topk: {len(hits)}")
        # for hit in hits:
        #     print(hit.to_dict())

    print(f"============== time cost {time.time()-start} seconds")


for i in range(10):
    search({'params': {}, 'metric_type': metric_type})
```

The result:
```
v2.3.4
collection loaded
count(*): 1000000
collection.num_entities =  1000000
=================== search ==================
============== time cost 0.44342780113220215 seconds
=================== search ==================
============== time cost 0.327681303024292 seconds
=================== search ==================
============== time cost 0.3355672359466553 seconds
=================== search ==================
============== time cost 0.27884340286254883 seconds
=================== search ==================
============== time cost 0.3779788017272949 seconds
=================== search ==================
============== time cost 0.31079530715942383 seconds
=================== search ==================
============== time cost 0.3096804618835449 seconds
=================== search ==================
============== time cost 0.34488368034362793 seconds
=================== search ==================
============== time cost 0.293459415435791 seconds
=================== search ==================
============== time cost 0.4231123924255371 seconds

Process finished with exit code 0

```

**Reply by YangZhaoo** [reply]
What is your machine configuration like?
I use this test script to connect remote milvus, It takes about 800ms.
when I switch metric type from L2 to COSINE,  It takes about 1200ms.
Is cosine similarity more time consuming?

**Reply by yhmo** [reply]
My milvus is a standalone with default configurations deployed by docker-compose on a machine with 3.2GHz CPU 12 cores and 64GB RAM.

Switch metric type from L2 to COSINE, the performance is no much difference.
```
=================== search ==================
============== time cost 0.43256497383117676 seconds
=================== search ==================
============== time cost 0.40921998023986816 seconds
=================== search ==================
============== time cost 0.3768901824951172 seconds
=================== search ==================
============== time cost 0.37567138671875 seconds
=================== search ==================
============== time cost 0.4354512691497803 seconds
=================== search ==================
============== time cost 0.34457993507385254 seconds
=================== search ==================
============== time cost 0.34348177909851074 seconds
=================== search ==================
============== time cost 0.36554789543151855 seconds
=================== search ==================
============== time cost 0.3651573657989502 seconds
=================== search ==================
============== time cost 0.3488180637359619 seconds
```

**Reply by YangZhaoo** [reply]
Are there other optimization methods? For example, modify the [configuration file](https://milvus.io/docs/configure-docker.md#Modify-the-configuration-file), [scale out or scale up query node](https://milvus.io/docs/configure-docker.md#Modify-the-configuration-file). Or, it can only be accelerated by GPU.
Thank you

**Reply by yhmo** [reply]
I just use default configurations. For FLAT index, no much optimization, it mainly depends on CPU ability.

GPU is suitable for large batch query, with nq>1000. 60 vectors query is not a large batch. If your CPU is not good, you can try GPU.

**Reply by YangZhaoo** [reply]
I ran your script to reduce the data size from 1,000,000 to 100,000. It was found that the latency increased instead. from 500ms to 900ms, I'd like to know what your results were. and why this happens.
Thank you.
`for i in range(100)  =====>>>     for i in range(10)`
`[f"name_{k%20}" for k in range(batch_size)] =====>>>  [f"name_{k%2}" for k in range(batch_size)]`
`expr="name like 'name_4%'",  =====>>>  expr="name like 'name_1%'",`

**Reply by yhmo** [reply]
- 1,000,000 rows
[f"name_{k%20}" for k in range(batch_size)]  --------------------------------------- 1/20 rows are named begin with "name_4"
expr="name like 'name_4%'"  ----------------------------------------------------------- 1/20 rows are filtered out, 50000 rows are computed


- 100,000 rows
[f"name_{k%2}" for k in range(batch_size)] ------------------------------------------ 1/2 rows are named begin with "name_1"
expr="name like 'name_1%'" ------------------------------------------------------------ 1/2 rows are filtered our, 50000 rows are computed


My results:
```
count(*): 100000
collection.num_entities =  100000
=================== search ==================
============== time cost 0.3576340675354004 seconds
=================== search ==================
============== time cost 0.17081427574157715 seconds
=================== search ==================
============== time cost 0.20328021049499512 seconds
=================== search ==================
============== time cost 0.2997429370880127 seconds
=================== search ==================
============== time cost 0.3802165985107422 seconds
=================== search ==================
============== time cost 0.27300214767456055 seconds
=================== search ==================
============== time cost 0.2394390106201172 seconds
=================== search ==================
============== time cost 0.16273760795593262 seconds
=================== search ==================
============== time cost 0.2594332695007324 seconds
=================== search ==================
============== time cost 0.2351069450378418 seconds
```

Average time cost 200~300ms, faster than 1000000 rows.
The time cost for computing 50000 rows is equal, but the filtering time cost of 1000000 is longer than 100000. So, the performance makes sense on my machine. Not sure why your machine is so slow. You can use two machines to compare.
